### PR TITLE
Documentation and Code Updates

### PR DESCRIPTION
**Pull request by jshier**
This PR updates AFI to build with the latest Alamofire code (for now), as well as updating the README to reflect the new API names and adding a Migration Guide for version 3.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -134,9 +138,9 @@ There are several `UIImage` extensions designed to make the common image manipul
 #### Inflation
 
 ```swift
-let URL = NSBundle.mainBundle().URLForResource("unicorn", withExtension: "png")!
-let data = NSData(contentsOfURL: URL)!
-let image = UIImage(data: data, scale: UIScreen.mainScreen().scale)!
+let url = Bundle.main.url(forResource: "unicorn", withExtension: "png")!
+let data = (try? Data(contentsOf: url))!
```
Comment (cnoon) [inline README.md]: `let data = try! Data(contentsOf: url)!` instead?

File: README.md [hunk 2]
```diff
@@ -174,11 +178,9 @@ let circularImage = image.af_imageRoundedIntoCircle()
...
 let image = UIImage(named: "unicorn")!
 
-let sepiaImage = image.af_imageWithAppliedCoreImageFilter("CISepiaTone")
-let blurredImage = image.af_imageWithAppliedCoreImageFilter(
-    "CIGaussianBlur", 
-    filterParameters: ["inputRadius": 25]
-)
+let sepiaImage = image.af_imageFiltered(withCoreImageFilter: "CISepiaTone")
+let blurredImage = image.af_imageFiltered(withCoreImageFilter: "CIGuassianBlue",
+                                          parameters: ["inputRadius": 25])
```
Comment (cnoon) [inline README.md]: Should be:

```swift
let blurredImage = image.af_imageFiltered(
    withCoreImageFilter: "CIGuassianBlue",
    parameters: ["inputRadius": 25]
)
```

Comment (jshier) [inline README.md]: Xcode now does this indentation automatically when breaking lines up, you want this other style anyway?

File: README.md [hunk 3]
```diff
@@ -237,13 +239,13 @@ The current list of multi-pass image filters includes:
...
+For example, let's say you need to download an album of images. Your app needs to display both the thumbnail version as well as the full size version at various times. Due to performance issues, you want to scale down the thumbnails to a reasonable size before rendering them on-screen. You also need to apply a global CoreImage filter to the full size images when displayed. While `URLCache` can easily handle storing the original downloaded image, it cannot store these different variants. What you really need is another caching layer designed to handle these different variants.
 
 ```swift
 let imageCache = AutoPurgingImageCache(
-    memoryCapacity: 100 * 1024 * 1024, 
+    memoryCapacity: 100 * 1024 * 1024,
     preferredMemoryUsageAfterPurge: 60 * 1024 * 1024
 )
```
Comment (cnoon) [inline README.md]: These should probably match the formatting that I switched over to in the initializers.

```swift
 let imageCache = AutoPurgingImageCache(
    memoryCapacity: 100_000_000,
    preferredMemoryUsageAfterPurge: 60_000_000
  )
```

File: README.md [hunk 4]
```diff
@@ -259,55 +261,42 @@ let imageCache = AutoPurgingImageCache()
...
 
 // Remove
-imageCache.removeImageWithIdentifier("avatar")
+imageCache.removeImage(withIdentifier: "avatar")
 ```
 
 #### URL Requests
 
-The `ImageRequestCache` protocol extends the `ImageCache` protocol by adding support for `NSURLRequest` caching. This allows an `NSURLRequest` and additional identifier to generate the unique identifier for the image in the cache.
+The `ImageRequestCache` protocol extends the `ImageCache` protocol by adding support for `URLRequest` caching. This allows an `URLRequest` and an additional identifier to generate the unique identifier for the image in the cache.
```
Comment (cnoon) [inline README.md]: `This allows a `URLRequest` and an additional...

File: README.md [hunk 5]
```diff
@@ -470,19 +459,15 @@ If an image filter is specified, it is applied asynchronously after the remote i
...
-let URL = NSURL(string: "https://httpbin.org/image/png")!
+let url = URL(string: "https://httpbin.org/image/png")!
 let placeholderImage = UIImage(named: "placeholder")!
 
-let filter = AspectScaledToFillSizeWithRoundedCornersFilter(
-    size: imageView.frame.size, 
-    radius: 20.0
-)
+let filter = AspectScaledToFillSizeWithRoundedCornersFilter(size: imageView.frame.size,
+                                                            radius: 20.0)
```
Comment (cnoon) [inline README.md]: Should probably be:

```swift
let filter = AspectScaledToFillSizeWithRoundedCornersFilter(
    size: imageView.frame.size,
    radius: 20.0
)
```

File: README.md [hunk 6]
```diff
@@ -494,20 +479,16 @@ By default, there is no image transition animation when setting the image on the
...
```
Comment (cnoon) [inline README.md]: Should probably be:

```swift
let filter = AspectScaledToFillSizeWithRoundedCornersFilter(
    size: imageView.frame.size,
    radius: 20.0
)
```

File: README.md [hunk 7]
```diff
@@ -494,20 +479,16 @@ By default, there is no image transition animation when setting the image on the
...
-imageView.af_setImageWithURL(
-    URL, 
-    placeholderImage: placeholderImage,
-    filter: filter,
-    imageTransition: .CrossDissolve(0.2)
-)
+imageView.af_setImage(withURL: url,
+                      placeholderImage: placeholderImage,
+                      filter: filter,
+                      imageTransition: .crossDissolve(0.2))
```
Comment (cnoon) [inline README.md]: Should probably be:

```swift
imageView.af_setImage(
    withURL: url,
    placeholderImage: placeholderImage,
    filter: filter,
    imageTransition: .crossDissolve(0.2)
)
```

File: Source/UIButton+AlamofireImage.swift [hunk 8]
```diff
@@ -408,8 +409,9 @@ extension UIButton {
         -> Bool
     {
         if
-            let currentRequest = backgroundImageRequestReceipt(for: state)?.request.task.originalRequest,
-            currentRequest.urlString == urlRequest?.urlRequest.urlString
+            let currentRequestURL = backgroundImageRequestReceipt(for: state)?.request.task.originalRequest?.url,
+            let requestURL = urlRequest?.urlRequest.url,
+            currentRequestURL.absoluteString == requestURL.absoluteString
```
Comment (cnoon) [inline Source/UIButton+AlamofireImage.swift]: `currentRequestURL == requestURL`?

File: Source/UIImageView+AlamofireImage.swift [hunk 9]
```diff
@@ -372,8 +372,9 @@ extension UIImageView {
 
     private func isURLRequestURLEqualToActiveRequestURL(_ urlRequest: URLRequestConvertible?) -> Bool {
         if
-            let currentRequest = af_activeRequestReceipt?.request.task.originalRequest,
-            currentRequest.urlString == urlRequest?.urlRequest.urlString
+            let currentRequestURL = af_activeRequestReceipt?.request.task.originalRequest?.url,
+            let requestURL = urlRequest?.urlRequest.url,
+            currentRequestURL.absoluteString == requestURL.absoluteString
```
Comment (cnoon) [inline Source/UIImageView+AlamofireImage.swift]: `currentRequestURL == requestURL`?

File: Tests/ImageDownloaderTests.swift [hunk 10]
```diff
@@ -921,7 +921,7 @@ class ImageDownloaderTestCase: BaseTestCase {
         // Given
         let downloader = ImageDownloader()
         let urlRequest = URLRequest(urlString: "https://httpbin.org/image/jpeg", method: .get)
-        let request = downloader.sessionManager.request(urlRequest)
+        let request = downloader.sessionManager.request(resource: urlRequest)
```
Comment (cnoon) [inline Tests/ImageDownloaderTests.swift]: Can now revert.

File: Tests/ImageDownloaderTests.swift [hunk 11]
```diff
@@ -942,8 +942,8 @@ class ImageDownloaderTestCase: BaseTestCase {
         let urlRequest1 = URLRequest(urlString: "https://httpbin.org/image/jpeg", method: .get)
         let urlRequest2 = URLRequest(urlString: "https://httpbin.org/image/png", method: .get)
 
-        let request1 = downloader.sessionManager.request(urlRequest1)
-        let request2 = downloader.sessionManager.request(urlRequest2)
+        let request1 = downloader.sessionManager.request(resource: urlRequest1)
+        let request2 = downloader.sessionManager.request(resource: urlRequest2)
```
Comment (cnoon) [inline Tests/ImageDownloaderTests.swift]: Can now revert.

File: Tests/ImageDownloaderTests.swift [hunk 12]
```diff
@@ -964,8 +964,8 @@ class ImageDownloaderTestCase: BaseTestCase {
...
```
Comment (cnoon) [inline Tests/ImageDownloaderTests.swift]: Can now revert.

File: Tests/ImageDownloaderTests.swift [hunk 13]
```diff
@@ -986,8 +986,8 @@ class ImageDownloaderTestCase: BaseTestCase {
...
```
Comment (cnoon) [inline Tests/ImageDownloaderTests.swift]: Can now revert.

**Comment by cnoon** [comment]
Made it through...looks great @jshier! Just a few minor tweaks and formatting updates.

File: README.md (hunk 1 above)
Comment (jshier) [inline README.md]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: README.md (hunk 2 above)
Comment (jshier) [inline README.md]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: README.md (hunk 3 above)
Comment (jshier) [inline README.md]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: README.md (hunk 5 above)
Comment (jshier) [inline README.md]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: README.md (hunk 6 above)
Comment (jshier) [inline README.md]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: README.md (hunk 7 above)
Comment (jshier) [inline README.md]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: Tests/ImageDownloaderTests.swift (hunk 10 above)
Comment (jshier) [inline Tests/ImageDownloaderTests.swift]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: Tests/ImageDownloaderTests.swift (hunk 11 above)
Comment (jshier) [inline Tests/ImageDownloaderTests.swift]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: Tests/ImageDownloaderTests.swift (hunk 12 above)
Comment (jshier) [inline Tests/ImageDownloaderTests.swift]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: Tests/ImageDownloaderTests.swift (hunk 13 above)
Comment (jshier) [inline Tests/ImageDownloaderTests.swift]: Addressed in [ec8d30](https://github.com/Alamofire/AlamofireImage/pull/176/commits/ec8d3083d871b7aa6d64670f2172e16231d6f8ab).

File: README.md (hunk 4 above)
Comment (jshier) [inline README.md]: Addressed in [79cf65](https://github.com/Alamofire/AlamofireImage/pull/176/commits/79cf65df27791b3a53c9ad13a791879407eb26dd).

File: Source/UIButton+AlamofireImage.swift (hunk 8 above)
Comment (jshier) [inline Source/UIButton+AlamofireImage.swift]: Addressed in [0e0ce](https://github.com/Alamofire/AlamofireImage/pull/176/commits/0e0ce9cb44eaf1147aca546799caf54de76121fd).

File: Source/UIImageView+AlamofireImage.swift (hunk 9 above)
Comment (jshier) [inline Source/UIImageView+AlamofireImage.swift]: Addressed in [0e0ce](https://github.com/Alamofire/AlamofireImage/pull/176/commits/0e0ce9cb44eaf1147aca546799caf54de76121fd).

**Comment by jshier** [comment]
@cnoon Comments addressed.

**Pull request merged**